File: src/codegen.rs
use inkwell::OptimizationLevel;
use inkwell::context::Context;

use crate::ast::Program;
use crate::errors::CodegenError;

/// Generate LLVM IR from a parsed [`Program`] and write the resulting
/// object file (or execute it, depending on future CLI flags).
///
/// Currently this sets up an LLVM module and does nothing with the AST
/// since no language constructs exist yet.
pub fn generate(_program: &Program) -> Result<(), CodegenError> {
    let context = Context::create();
    let module = context.create_module("welang");
    let _execution_engine = module
        .create_jit_execution_engine(OptimizationLevel::None)
        .map_err(|e| CodegenError::LlvmError {
            message: e.to_string(),
        })?;

    // TODO: walk the AST and emit LLVM IR.

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ast::Program;

    #[test]
    fn generate_empty_program() {
        let program = Program { items: vec![] };
        let result = generate(&program);
        assert!(
            result.is_ok(),
            "codegen on empty program failed: {:?}",
            result
        );
    }

    #[test]
    fn llvm_context_creation() {
        // Verify we can create an LLVM context without panicking.
        let _ctx = Context::create();
    }

    #[test]
    fn llvm_module_creation() {
        let ctx = Context::create();
        let module = ctx.create_module("test");
        assert_eq!(module.get_name().to_str().unwrap(), "test");
    }
}

File: Sources/WeLangLib/Codegen.swift
import CLLLVM

/// Generate LLVM IR from a parsed `Program` and write the resulting
/// object file (or execute it, depending on future CLI flags).
///
/// Currently this sets up an LLVM module and does nothing with the AST
/// since no language constructs exist yet.
public func generate(_ program: Program) throws {
    let context = LLVMContextCreate()
    defer { LLVMContextDispose(context) }

    let module = LLVMModuleCreateWithNameInContext("welang", context)
    defer { LLVMDisposeModule(module) }

    // Initialize the native target so the JIT can work.
    LLVMLinkInMCJIT()
    LLVM_InitializeNativeTarget()
    LLVM_InitializeNativeAsmPrinter()

    var executionEngine: LLVMExecutionEngineRef?
    var errorMessage: UnsafeMutablePointer<CChar>?

    let result = LLVMCreateJITCompilerForModule(&executionEngine, module, 0, &errorMessage)
    if result != 0 {
        let message: String
        if let errorMessage {
            message = String(cString: errorMessage)
            LLVMDisposeMessage(errorMessage)
        } else {
            message = "unknown LLVM error"
        }
        throw CodegenError.llvmError(message: message)
    }

    defer {
        if let executionEngine {
            LLVMDisposeExecutionEngine(executionEngine)
        }
    }

    // TODO: walk the AST and emit LLVM IR.
}

// MARK: - LLVM Helpers

/// Creates an LLVM context. Caller is responsible for disposal.
public func createLLVMContext() -> LLVMContextRef? {
    return LLVMContextCreate()
}

/// Creates an LLVM module in the given context. Caller is responsible for disposal.
public func createLLVMModule(named name: String, in context: LLVMContextRef?) -> LLVMModuleRef? {
    return LLVMModuleCreateWithNameInContext(name, context)
}

/// Returns the name of an LLVM module as a String.
public func getLLVMModuleName(_ module: LLVMModuleRef?) -> String? {
    guard let module else { return nil }
    var length: Int = 0
    guard let cStr = LLVMGetModuleIdentifier(module, &length) else { return nil }
    return String(cString: cStr)
}

// MARK: - Native Target Initialization Helpers

private func LLVM_InitializeNativeTarget() {
    LLVMInitializeX86TargetInfo()
    LLVMInitializeX86Target()
    LLVMInitializeX86TargetMC()
}

private func LLVM_InitializeNativeAsmPrinter() {
    LLVMInitializeX86AsmPrinter()
}
